File: bot/src/gpt/tool/function/SendLaterFunction.ts
import AssistantFunction, {AssistantObject, InvocationContext} from "./AssistantFunction";
import {VkMessage} from "../../../service/VkMessagesService";
import {Context} from "../../../Context";
import {Logger} from "../../../service/LoggingService";

export default class SendLaterFunction implements AssistantFunction {
    private logger!: Logger;

    constructor(private context: Context) {
        context.onReady(() => {
            this.logger = context.loggingService.getRootLogger().newSublogger("SendLaterFunction");
        });
    }

    getDescription(): string {
        return "Sends a message later";
    }

    getName(): string {
        return "send_later";
    }

    getParameters(): AssistantObject {
        return {
            type: "object",
            properties: {
                delay: {
                    type: "integer",
                    description: "Delay in seconds"
                },
                message: {
                    type: "string",
                    description: "Message to send"
                }
            },
            required: ["delay", "message"]
        };
    }

    async call(args: any, message: VkMessage, invocationContext: InvocationContext): Promise<string> {
        const messageText = args['message'];
        const waitSeconds = args['delay'];
        if (isNaN(waitSeconds)) {
            throw new Error("Delay should be a number");
        }
        if (!messageText) {
            throw new Error("Message should not be empty");
        }
        if (waitSeconds <= 0) {
            throw new Error("Delay should be positive");
        }
        await this.context.deferredVkMessagesService.sendLater(message.peerId, messageText, waitSeconds);
        this.logger.newSublogger(`peer_id:${message.peerId}`).info(
            `Scheduled message to be sent after ${waitSeconds} seconds`
        );
        return "Message is scheduled to be sent later";
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/gpt/tool/function/SendLaterFunction.kt
package space.davids_digital.sweetie.gpt.tool.function

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import space.davids_digital.sweetie.gpt.InvocationContext
import space.davids_digital.sweetie.gpt.tool.function.parameter.Description
import space.davids_digital.sweetie.model.VkMessageModel
import space.davids_digital.sweetie.service.DeferredVkMessageService

@Component
class SendLaterFunction(
    private val deferredVkMessageService: DeferredVkMessageService
): AssistantFunction<SendLaterFunction.SendLaterParameters> {
    companion object {
        private val log = LoggerFactory.getLogger(SendLaterFunction::class.java)
    }

    data class SendLaterParameters(
        @Description("Message to send")
        val message: String,
        @Description("Delay in seconds")
        val delaySeconds: Long
    )

    override fun getName() = "schedule_message"
    override fun getDescription() = "Send a message later"
    override fun getParametersClass() = SendLaterParameters::class

    override suspend fun call(
        parameters: SendLaterParameters,
        message: VkMessageModel,
        invocationContext: InvocationContext
    ): String {
        if (parameters.message.isBlank()) {
            return "Message should not be empty"
        }
        if (parameters.delaySeconds <= 0) {
            return "Delay should be greater than 0"
        }
        deferredVkMessageService.scheduleMessage(message.peerId, parameters.message, parameters.delaySeconds)
        log.info("Scheduled message to ${message.peerId} in ${parameters.delaySeconds} seconds")
        return "Message scheduled"
    }
}